soccer/radio/SimRadio.cpp

```diff
@@ -79,68 +85,80 @@ void SimRadio::send(Packet::RadioTx& radioTx) {
-        _rx_socket.readDatagram(&byte, 1);
-
-        // grSim really needs to set up a proto packet for robot status.
-        // Instead they pack their own byte up in a custom way :/
-        //
-        // byte structure:
-        // 0-2: Robot_ID
-        // 3: touching_ball
-        // 4: just_kicked
-        // 5: robot_on
-        const uint8_t robot_id_mask = 0x7;
-        const uint8_t touching_ball_mask = 0x1 << 3;
-        const uint8_t just_kicked_mask = 0x1 << 4;
-        const uint8_t robot_on_mask = 0x1 << 5;
-
-        int robot_id = byte & robot_id_mask;
-        bool ball_sense = byte & touching_ball_mask;
-        bool just_kicked = byte & just_kicked_mask;
-        bool robot_on = byte & robot_on_mask;
-
-        RadioRx rx;
-        rx.set_robot_id(robot_id);
-        rx.set_hardware_version(RJ2015);
-        rx.set_battery(100);
-
-        rx.set_ball_sense_status(ball_sense ? Packet::HasBall : Packet::NoBall);
-
-        const int num_motors = 5;
-        // 5 motors including dribbler
-        for (int i = 0; i < num_motors; i++) {
-            rx.add_motor_status(MotorStatus::Good);
-        }
+    _socket.send_to(buffer(out), _grsim_endpoint);
+}
 
-        rx.set_fpga_status(FpgaGood);
-        rx.set_timestamp(RJ::timestamp());
+void SimRadio::receive() { _io_service.poll(); }
 
-        const uint8_t kicker_status_charging = Kicker_Enabled | Kicker_I2C_OK;
-        const uint8_t kicker_status_ready =
-            Kicker_Charged | kicker_status_charging;
-        ;
+void SimRadio::startReceive() {
+    // Set a receive callback
+    _socket.async_receive(
+        boost::asio::buffer(_buffer),
+        [this](const boost::system::error_code& error, std::size_t num_bytes) {
+            receivePacket(error, num_bytes);
+        });
+}
 
-        rx.set_kicker_status(just_kicked ? kicker_status_charging
-                                         : kicker_status_ready);
-        rx.set_kicker_voltage(200);
+void SimRadio::receivePacket(const boost::system::error_code& error,
+                             std::size_t num_bytes) {
+    for (int i = 0; i < num_bytes; i++) {
+        handleReceive(_buffer[i]);
+    }
+    startReceive();
+}
 
-        std::lock_guard<std::mutex> lock(_reverse_packets_mutex);
-        _reversePackets.push_back(rx);
+void SimRadio::handleReceive(uint8_t data) {
+    // grSim really needs to set up a proto packet for robot status.
+    // Instead they pack their own byte up in a custom way :/
+    //
+    // byte structure:
+    // 0-2: Robot_ID
+    // 3: touching_ball
+    // 4: just_kicked
+    // 5: robot_on
+    const uint8_t robot_id_mask = 0x7;
+    const uint8_t touching_ball_mask = 0x1 << 3;
+    const uint8_t just_kicked_mask = 0x1 << 4;
+    const uint8_t robot_on_mask = 0x1 << 5;
+
+    int this_robot_id = data & robot_id_mask;
+    bool ball_sense = data & touching_ball_mask;
+    bool just_kicked = data & just_kicked_mask;
+    bool robot_on = data & robot_on_mask;
```

Comment:
for clarity I think you could separate the parsing logic out into a different function and return a struct that contains all the information that a packet could contain

Reply:
This function is about to be rewritten (grSim update introduces a new protocol) so I don't really want to bother until it's actually being updated.